File: react/src/api/canvas.ts
import { CanvasData, Message, Session } from '@/types/types'
import { ToolInfo } from '@/api/model'

export type ListCanvasesResponse = {
  id: string
  name: string
  description?: string
  thumbnail?: string
  created_at: string
}

export async function listCanvases(): Promise<ListCanvasesResponse[]> {
  const response = await fetch('/api/canvas/list')
  return await response.json()
}

export async function createCanvas(data: {
  name: string
  canvas_id: string
  messages: Message[]
  session_id: string
  text_model: {
    provider: string
    model: string
    url: string
  }
  tool_list: ToolInfo[]

  system_prompt: string
}): Promise<{ id: string }> {
  const response = await fetch('/api/canvas/create', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  })
  return await response.json()
}

export async function getCanvas(
  id: string
): Promise<{ data: CanvasData | null; name: string; sessions: Session[] }> {
  try {
    const response = await fetch(`/api/canvas/${id}`)

    if (!response.ok) {
      console.warn(`Canvas API returned ${response.status}: ${response.statusText}`)
      // 返回默认数据而不是抛出错误
      return {
        data: null,
        name: '未命名画布',
        sessions: []
      }
    }

    const data = await response.json()

    // 更宽松的数据验证
    if (!data) {
      console.warn('Canvas API returned null/undefined data')
      return {
        data: null,
        name: '未命名画布',
        sessions: []
      }
    }

    // 确保返回的数据结构正确
    return {
      data: data.data || null,
      name: data.name || '未命名画布',
      sessions: Array.isArray(data.sessions) ? data.sessions : []
    }
  } catch (error) {
    console.error('Error fetching canvas:', error)
    // 网络错误或其他异常时返回默认数据
    return {
      data: null,
      name: '未命名画布',
      sessions: []
    }
  }
}

export async function saveCanvas(
  id: string,
  payload: {
    data: CanvasData
    thumbnail: string
  }
): Promise<void> {
  const response = await fetch(`/api/canvas/${id}/save`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (!response.ok) {
    throw new Error(`Failed to save canvas: ${response.status} ${response.statusText}`)
  }
}

export async function renameCanvas(id: string, name: string): Promise<void> {
  const response = await fetch(`/api/canvas/${id}/rename`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  })
  return await response.json()
}

export async function deleteCanvas(id: string): Promise<void> {
  const response = await fetch(`/api/canvas/${id}/delete`, {
    method: 'DELETE',
  })
  return await response.json()
}

File: server/routers/canvas.py
from fastapi import APIRouter, Request, HTTPException
from starlette.requests import ClientDisconnect
#from routers.agent import chat
from services.chat_service import handle_chat
from services.db_service import db_service
import asyncio
import json

router = APIRouter(prefix="/api/canvas")

@router.get("/list")
async def list_canvases():
    return await db_service.list_canvases()

@router.post("/create")
async def create_canvas(request: Request):
    data = await request.json()
    id = data.get('canvas_id')
    name = data.get('name')

    asyncio.create_task(handle_chat(data))
    await db_service.create_canvas(id, name)
    return {"id": id }

@router.get("/{id}")
async def get_canvas(id: str):
    return await db_service.get_canvas_data(id)

@router.post("/{id}/save")
async def save_canvas(id: str, request: Request):
    try:
        payload = await request.json()
        data_str = json.dumps(payload['data'])
        await db_service.save_canvas_data(id, data_str, payload['thumbnail'])
        return {"id": id }
    except ClientDisconnect:
        # 客户端断开连接（通常发生在自动保存时用户关闭页面）
        print(f"⚠️ 客户端断开连接，画布 {id} 自动保存被中断")
        # 不抛出异常，静默处理
        return {"id": id, "status": "partial"}
    except Exception as e:
        # 记录详细错误信息
        import traceback
        print(f"❌ 保存画布 {id} 失败:")
        print(f"   错误类型: {type(e).__name__}")
        print(f"   错误信息: {str(e)}")
        print(f"   堆栈跟踪:")
        traceback.print_exc()
        
        # 返回友好的错误响应
        raise HTTPException(status_code=500, detail=f"保存失败: {str(e)}")

@router.post("/{id}/rename")
async def rename_canvas(id: str, request: Request):
    data = await request.json()
    name = data.get('name')
    await db_service.rename_canvas(id, name)
    return {"id": id }

@router.delete("/{id}/delete")
async def delete_canvas(id: str):
    await db_service.delete_canvas(id)
    return {"id": id }